src/core/broker_config.cpp

```diff
@@ -17,80 +17,282 @@
+  }
+};
+// clang-format on
+
+// Read the JSON configuration from a file.
+BrokerConfig::ParseResult BrokerConfig::Read(const std::string& file_name)
 {
   std::ifstream file_stream(file_name);
   if (!file_stream.is_open())
     return ParseResult::kFileOpenErr;
 
-  return ReadFromStream(file_stream);
+  return Read(file_stream);
 }
 
-BrokerConfig::ParseResult BrokerConfig::ReadFromStream(std::istream& stream)
+// Read the JSON configuration from an input stream.
+BrokerConfig::ParseResult BrokerConfig::Read(std::istream& stream)
 {
   try
   {
     stream >> current_;
-    return Validate(current_, defaults_);
+    return ValidateCurrent();
   }
   catch (json::parse_error)
   {
     return ParseResult::kJsonParseErr;
   }
 }
 
-BrokerConfig::ParseResult BrokerConfig::Validate(const json& current_obj, const json& default_obj)
+// Validate the JSON object contained in the "current_" member, which presumably has just been
+// deserialized. Currently, extra keys present in the JSON which we don't recognize are considered
+// valid.
+BrokerConfig::ParseResult BrokerConfig::ValidateCurrent()
 {
-  for (auto item = current_obj.begin(); item != current_obj.end(); ++item)
+  for (auto setting : kSettingsValidatorArray)
   {
-    // Check each key that matches an item in our default settings object.
-    if (default_obj.contains(item.key()))
+    try
     {
-      if (default_obj[item.key()].type() != item.value().type())
+      // Check each key that matches an item in our settings array.
+      if (current_.contains(setting.pointer))
       {
-        // The value type of this item does not match the type of the corresponding default setting.
-        return ParseResult::kInvalidSetting;
-      }
+        json val = current_[setting.pointer];
+
+        // If a setting is set to "null" in the JSON, store the default value.
+        if (val.is_null() && setting.store_default)
+        {
+          setting.store_default(*this);
+        }
+        else
+        {
+          if (val.type() != setting.type)
+          {
+            // The value type of this item does not match the type of the corresponding default setting.
+            return ParseResult::kInvalidSetting;
+          }
 
-      // Recurse into sub-objects.
-      if (item.value().is_object())
+          // Try to validate the setting's value.
+          if (!setting.validate_and_store(val, *this))
+          {
+            return ParseResult::kInvalidSetting;
+          }
+        }
+      }
+      else if (setting.store_default)
       {
-        ParseResult sub_result = Validate(item.value(), default_obj[item.key()]);
-        if (sub_result != ParseResult::kOk)
-          return sub_result;
+        // The "store_default" function may not be present, in which case the default-constructed
+        // value in the output settings is left as-is.
+        setting.store_default(*this);
       }
     }
+    catch (json::parse_error)
+    {
+      return ParseResult::kInvalidSetting;
```

Comment:
Should this be ParseResult::kJsonParseErr?

Reply:
Yeah, I was trying to differentiate between overall poorly-formed JSON and a poorly formed value for a key. But I think I bungled it. I will revisit all the parse_error stuff.